### PR TITLE
PG policy

### PR DESCRIPTION
**Pull request by norci**
for #80 .

* I implemented code for discrete action space. The other kinds of space are not implemented yet.

* I'm not familiar with Julia's code style. Is this code OK?

* I referenced code from [backwardpass.jl](https://github.com/kraftpunk97/Monte-Carlo-PG-in-julia/blob/768c519cacf945d256e71f36376a6006da8ecfb1/backwardpass.jl#L9), it normalized discounted rewards.
This method is able to improve the training process.

image
The blue line is normalized discounted reward version.
The red line is standard version.

* can we distinguish envs by action & observation space by their types?
I think this style is ugly, I wish it can be implemented in a C++ template way.
```
    actions = get_actions(env)
    if typeof(actions) <: DiscreteSpace
        action = actions[rand(dist)]
    elseif typeof(actions) <: ContinuousSpace
        action = clamp(rand(dist), actions.low, actions.high)
```

Also in the update! function, how can we know if the action space is continuous or discrete?

### CODE REVIEW

File: src/algorithms/policy_gradient/pg.jl [hunk 1]
```diff
@@ -0,0 +1,63 @@
...
+    approximator::A
+    dist_fn::Any = Normal # distribution function
+    γ::Float32 = 0.99f0 # discount factor
+    rng::R = Random.GLOBAL_RNG
+    loss::Float32 = 0f0
+end
+
+# if the action space is discrete, then softmax should be the last layer of the model.
+
+function (π::PGPolicy)(env)
```
Comment (findmyway) [inline src/algorithms/policy_gradient/pg.jl]: You can leverage multiple-dispatch here.

```julia
(π::PGPolicy)(env) = π(env, get_actions(env))

function (π::PGPolicy)(env, actions)
    # the default implementation assumes actions are discrete
    # do sth
end

function (π::PGPolicy)(env, actions::ContinuousSpace)
    # implementation specific to continuous actions
end

File: src/algorithms/policy_gradient/pg.jl [hunk 2]
```diff
@@ -0,0 +1,63 @@
...
+Base.@kwdef mutable struct PGPolicy{A<:NeuralNetworkApproximator,R<:AbstractRNG} <:
+                           AbstractPolicy
+    approximator::A
+    dist_fn::Any = Normal # distribution function
+    γ::Float32 = 0.99f0 # discount factor
+    rng::R = Random.GLOBAL_RNG
+    loss::Float32 = 0f0
+end
+
+# if the action space is discrete, then softmax should be the last layer of the model.
```
Comment (findmyway) [inline src/algorithms/policy_gradient/pg.jl]: Better to move this part into the doc string part of `PGPolicy`

File: src/algorithms/policy_gradient/pg.jl [hunk 3]
```diff
@@ -0,0 +1,63 @@
...
+
+function (π::PGPolicy)(env)
+    to_dev = x -> send_to_device(device(π.approximator), x)
+
+    logits = env |> get_state |> to_dev |> π.approximator |> send_to_host # if the action space is discrete, then the approximator should softmax the output values.
+    dist = logits |> π.dist_fn
+
+    actions = get_actions(env)
+    if typeof(actions) <: DiscreteSpace
+        action = actions[rand(dist)]
```
Comment (findmyway) [inline src/algorithms/policy_gradient/pg.jl]:
```suggestion
        action = actions[rand(π.rng, dist)]
```

File: src/algorithms/policy_gradient/pg.jl [hunk 4]
```diff
@@ -0,0 +1,63 @@
...
+function RLBase.update!(π::PGPolicy, traj::ElasticCompactSARTSATrajectory)
+    (length(traj[:terminal]) > 0 && traj[:terminal][end]) || return
+
+    Q = π.approximator
+    D = device(Q)
+    to_dev = x -> send_to_device(D, x)
+
+    states = traj[:state] |> to_dev
+    actions = traj[:action] |> Array
+    rewards = traj[:reward] |> Array
```
Comment (findmyway) [inline src/algorithms/policy_gradient/pg.jl]: Why do we need to convert the view into `Array` here?

File: src/algorithms/policy_gradient/pg.jl [hunk 5]
```diff
@@ -0,0 +1,63 @@
...
+
+    gains = discount_rewards(rewards, π.γ) |> to_dev |> x -> Flux.normalise(x; dims = 1)
+
+    gs = gradient(Flux.params(Q)) do
+        logits = states |> Q |> Array
```
Comment (findmyway) [inline src/algorithms/policy_gradient/pg.jl]: Why do we need to convert `logits` into `Array` here?

File: src/algorithms/policy_gradient/pg.jl [hunk 6]
```diff
@@ -0,0 +1,63 @@
...
+    rng::R = Random.GLOBAL_RNG
+    loss::Float32 = 0f0
+end
+
+# if the action space is discrete, then softmax should be the last layer of the model.
+
+function (π::PGPolicy)(env)
+    to_dev = x -> send_to_device(device(π.approximator), x)
+
+    logits = env |> get_state |> to_dev |> π.approximator |> send_to_host # if the action space is discrete, then the approximator should softmax the output values.
```
Comment (findmyway) [inline src/algorithms/policy_gradient/pg.jl]: I think we can remove the `softmax` constraint of `approximator` here. So that we can use `logsoftmax` when updating to avoid some strange corner cases. If so, the test case needs to be updated.

**Review by findmyway — CHANGES_REQUESTED** [review]

**Comment by findmyway** [comment]
> I implemented code for discrete action space. The other kinds of space are not implemented yet.

That's fine. We can add a test case on Pendulum later.

> I'm not familiar with Julia's code style. Is this code OK?

LGTM

> I referenced code from backwardpass.jl, it normalized discounted rewards.

Cool. Maybe we can also add it to other PG implementations.

> can we distinguish envs by action & observation space by their types?

We can use multiple dispatch here. Replied in the review comments.

> Also in the update! function, how can we know if the action space is continuous or discrete?

Just use the same trick as above.

<hr>

I'm still on vacation, so my reply may be late.

**Comment by findmyway** [comment]
One thing I forgot to mention is that, we'd better add a `batch_size` in `PGPolicy`. In some environments, the episode length may be too large.

File: src/algorithms/policy_gradient/pg.jl (hunk 5 above)
Comment (norci) [reply src/algorithms/policy_gradient/pg.jl]: it's a mistake.

**Comment by norci** [comment]
@findmyway 

I implemented Vanilla Policy Gradient for continuous & discrete action space.
It can solve some discrete action space env, such as CartPole. Pendulum.
but it can not solve continuous action space Pendulum.
I'm not sure if there are bugs in the code, or VPG is not powerful enough.

**Comment by findmyway** [comment]
Thanks. I'll review the code after coming back from vacation tomorrow.

File: src/algorithms/policy_gradient/vpg.jl [hunk 7]
```diff
@@ -0,0 +1,156 @@
...
+    x = m.pre(S)
+    m.μ(x), m.σ(x) .|> exp
+end
+
+"""
+Vanilla Policy Gradient
+
+if the action space is continuous,
+then the env should transform the action value, (such as using tanh),
+in order to make sure low ≤ value ≤ high
```
Comment (findmyway) [inline src/algorithms/policy_gradient/vpg.jl:40]: It would be better to add a `# Keyword Arguments` section in the doc to give a short description of each argument.

File: src/algorithms/policy_gradient/vpg.jl [hunk 8]
```diff
@@ -0,0 +1,156 @@
...
+
+    model = π.approximator
+    to_dev(x) = send_to_device(device(model), x)
+
+    states = traj[:state]
+    actions = traj[:action]
+    gains = traj[:reward] |> x -> discount_rewards(x, π.γ)
+
+    for idx in Iterators.partition(shuffle(1:length(traj[:terminal])), π.batch_size)
+        S = states[:, idx] |> to_dev
```
Comment (findmyway) [inline src/algorithms/policy_gradient/vpg.jl]: Use `view` here to avoid unnecessary memory allocation. To support `states` of higher dimension, you're suggested to use the `select_last_dim` function provided in [`RLCore`](https://github.com/JuliaReinforcementLearning/ReinforcementLearningCore.jl/blob/a055fdad9324868d97fea7f7a31ba2b7a091051e/test/utils/base.jl#L14)

File: src/algorithms/policy_gradient/vpg.jl [hunk 9]
```diff
@@ -0,0 +1,156 @@
...
+        A = actions[idx]
+        G = gains[idx] |> to_dev |> x -> Flux.unsqueeze(x, 1)
+        # gains is a 1 colomn array, but the ouput of flux model is 1 row, n_batch columns array. so unsqueeze it.
+
+        if typeof(π.baseline) <: NeuralNetworkApproximator
```
Comment (findmyway) [inline src/algorithms/policy_gradient/vpg.jl]: `isa`

File: src/algorithms/policy_gradient/vpg.jl [hunk 10]
```diff
@@ -0,0 +1,156 @@
...
+                # TODO: is the loss function correct?
+                loss = mse(π.baseline(S), G) * π.α_w
+                ignore() do
+                    π.baseline_loss = loss
+                end
+                loss
+            end
+            update!(π.baseline, gs)
+        end
+        if typeof(π.baseline) <: Nothing
```
Comment (findmyway) [inline src/algorithms/policy_gradient/vpg.jl]: `isa`

Comment (findmyway) [reply src/algorithms/policy_gradient/vpg.jl]: I think that using `elseif` here would make the logic more clear?

File: src/experiments/rl_envs.jl [hunk 11]
```diff
@@ -1178,3 +1179,247 @@ function RLCore.Experiment(
...
+
+function RLCore.Experiment(
+    ::Val{:JuliaRL},
+    ::Val{:VPG},
+    ::Val{:Pendulum},
+    ::Nothing;
+    save_dir = nothing,
+    seed = 123,
+)
+    # TODO: vpg does not work.
```
Comment (findmyway) [inline src/experiments/rl_envs.jl]: How about removing this experiment for now given that it doesn't work yet?

**Review by findmyway — APPROVED** [review]

**Comment by findmyway** [comment]
By the way, you're encouraged to update https://github.com/JuliaReinforcementLearning/ReinforcementLearningZoo.jl#algorithms-implemented and https://github.com/JuliaReinforcementLearning/ReinforcementLearningZoo.jl#list-of-built-in-experiments

File: src/algorithms/policy_gradient/vpg.jl (hunk 8 above)
Comment (norci) [reply src/algorithms/policy_gradient/vpg.jl]: about `view`.
I found out this doc: https://docs.julialang.org/en/v1/manual/performance-tips/#Copying-data-is-not-always-bad

and did a benchmark:
`
julia> N=10000; X=randn(N); A=randn(N); inds = shuffle(1:N);
julia> @time sum(X.*A); @time sum(X[inds].*A[inds]); @time @views sum(X[inds].*A[inds]);
  0.000026 seconds (5 allocations: 78.281 KiB)
  0.000093 seconds (9 allocations: 234.688 KiB)
  0.000052 seconds (7 allocations: 78.469 KiB)
`

`@views` is faster than using slice directly.

I think the example in the julia doc did not count the time cost of memory allocation.

Comment (findmyway) [reply src/algorithms/policy_gradient/vpg.jl]: Thanks for the link.

[Copying-data-is-not-always-bad](https://docs.julialang.org/en/v1/manual/performance-tips/#Copying-data-is-not-always-bad) is correct. Usually, the algorithm will run on GPU and the data will be transformed into `Array` first:

https://github.com/JuliaGPU/CUDA.jl/blob/f31cbe22b4baba872a48bcb48e9f60e712f653fc/src/array.jl#L206

So there's no difference whether we use view or not in this case.

In CPU only environments I guess we can still add a *copy* layer to address the performance issue?

Comment (norci) [reply src/algorithms/policy_gradient/vpg.jl]: I benchmarked again, with larger 2d arrays.
using array slicing is much **faster** than `@views`.

```
julia> a=randn(1000,2000); b=randn(2000,1000);
julia> ind1 = shuffle(1:1000); ind2 = shuffle(1:2000);

julia> @time a*b;
  0.014062 seconds (2 allocations: 7.629 MiB)

julia> @time a[ind1,ind2]*b[ind2,ind1]; ## array slicing. fast, cost more memory
  0.027007 seconds (6 allocations: 38.147 MiB, 7.27% gc time)

julia> @time @views a[ind1,ind2]*b[ind2,ind1]; ## views. slow. does not require extra memory
  1.490200 seconds (14 allocations: 7.630 MiB)
```

I think in RL, the data size is very small usually, so there is no difference between slicing and `@views`.

For performance optimization, I think we should use a profiler to analyze the time cost of the code first.

what is a "copy layer"?
Can we simply use array slicing, or data |> Array?

```
julia> @time @views Array(a[ind1,ind2])*Array(b[ind2,ind1]);
  0.032226 seconds (12 allocations: 38.147 MiB)
```
Convert view to Array is still slower than array slicing, but much faster than using views.

Comment (findmyway) [reply src/algorithms/policy_gradient/vpg.jl]: > what is a "copy layer"?
> Can we simply use array slicing, or data |> Array?

Yes. That's what I mean here.

Comment (findmyway) [reply src/algorithms/policy_gradient/vpg.jl]: I think we can split this into another issue.

1. We stick with the `view` and users need to manually add an extra layer in `approximator` when working with CPU only . Just like

```
Chain(
    x -> convert(Array, x),
    # some other layers
)
```

2. Change the implementation of [`send_to_device`](https://github.com/JuliaReinforcementLearning/ReinforcementLearningCore.jl/blob/a94544cfceb4a6d527895414ead2744f66f23949/src/utils/device.jl#L12)

So we convert the `SubArray` from `view` to `Array` before feeding it into the approximator on CPU. Just like `Flux.cpu`.

I prefer the latter.

Comment (findmyway) [reply src/algorithms/policy_gradient/vpg.jl]: In any case, `states[:, idx]` still needs to be updated to handle states of higher dimensions.

Comment (findmyway) [reply src/algorithms/policy_gradient/vpg.jl]: Created an issue here: https://github.com/JuliaReinforcementLearning/ReinforcementLearningCore.jl/issues/130

File: src/experiments/rl_envs.jl (hunk 11 above)
Comment (norci) [reply src/experiments/rl_envs.jl]: for the continuous action space experiment, I suspect there are bugs in the algo, or I used wrong hyper parameter.
Let me check it carefully.

Comment (findmyway) [reply src/experiments/rl_envs.jl]: You are not alone. Haha https://discuss.pytorch.org/t/gym-pendulum-v0-not-solvable-by-vanilla-policy-gradient-increase-max-torques/8849/3

Let me know if you think it's ready to merge.

Comment (norci) [reply src/experiments/rl_envs.jl]: VPG can solve discrete action space Pendulum in 5 sec, and the performance is stable.
image

But for continuous action space, VPG is very difficult to converge.
image

the range of mean & std need to be selected carefully, or the agent can't learn anything.

and it needs a good seed.
I trained 12 agents.
6 agents can reach score -200.
3 agents can keep the average score > -300 in a few episodes.
but after that, the performance decreases gradually.
I guess it's the weakness of VPG, that it's variance is too large.

So I think there is no bug in the algo now.

Comment (rbange) [reply src/experiments/rl_envs.jl]: I just want to leave my small marks here:

1. Seeds should never be used to create good results. The reason this framework has a strong rng based implementation is to create reproducibility for single experiments (even though this is not guaranteed for major Julia updates). If the algorithm does only converge on a specific seed, then it probably needs some hyperparameter tuning somewhere else or is indeed unable to converge in this environment (as claimed [here](https://discuss.pytorch.org/t/gym-pendulum-v0-not-solvable-by-vanilla-policy-gradient-increase-max-torques/8849/3)) **The seed is no hyperparameter**

2. To be honest, Pendulum is also a bad environment to test and benchmark continuous control algorithms for 2 reasons: First, Pendulum performance is highly depended on the init values of the pendulum itself (>-100 performance seems only reachable if the pendulum spawns upright from the beginning). Second, pendulum only has a single dimension action space, which is often unrepresentative for real world control tasks, that's why you see Mujoco environments to benchmark the algorithms in most papers. I guess Pendulum is only used in this framework, because it is fairly easy to implement, fast and thereby very CI compatible.

I would therefore argue, that if you want to add this experiment, you should at least add a comment to the experiment which explains that this experimen only acts as a template implementation for the vpg in a continuous action space.

File: src/algorithms/policy_gradient/vpg.jl [hunk 12]
```diff
@@ -0,0 +1,140 @@
...
+            # normalise should not be used with baseline. or the loss of the policy will be too small.
+            δ = G |> x -> normalise(x; dims = 2)
+        end
+
+        gs = gradient(Flux.params(model)) do
+            if π.action_space isa DiscreteSpace
+                log_prob = S |> model |> logsoftmax
+                log_probₐ = log_prob[CartesianIndex.(A, 1:length(A))]
+            elseif π.action_space isa ContinuousSpace
+                dist = π.dist.(model(S)...) # TODO: Normal. does not work on GPU. InvalidIRError.
```
Comment (rbange) [inline src/algorithms/policy_gradient/vpg.jl]: Just took a glance at your implementation as this might be interesting for my project too.

I actually had the same issue with the Soft-Actor-Critic (the reason the implementation in this framework is not GPU capable). There seems to be an issue with drawing random numbers on the whole CUDA + Zygote stack. Unfortunatly, I was not able to really pinpoint which function was responsible for for the error.

However, I resolved this issue by implementing my own loglikelihood for normal distributions:

```
const log2π = log(2f0 * Float32(π))

function my_loglikelihood(μ, σ, x)
    z = (x .- μ) ./ σ
    -(z .^ 2 .+ log2π) ./ 2f0
end
```

Maybe this solves your issue too?

Edit:
Ah there's also this package: https://github.com/TuringLang/DistributionsAD.jl , which might resolve this issue by providing an adjoint for the `logpdf` function

Comment (norci) [reply src/algorithms/policy_gradient/vpg.jl]: I filed an issue for `Distributions`. https://github.com/JuliaStats/Distributions.jl/issues/1183
It seems the problem is related to CUDA, and can not be solved by using `DistributionsAD`.

I think your code is a better solution. Shall we put this code in somewhere, such as `ReinforcementLearningCore/src/utils/base.jl`?

I guess there is a typo in `log2π`?
```
function logpdf1(μ, σ, x)
    z = (x .- μ) ./ σ
    -(z .^ 2 .+ log(2π)) ./ 2f0
end
```

File: src/experiments/rl_envs.jl (hunk 11 above)
Comment (norci) [reply src/experiments/rl_envs.jl]: Thanks for your advice.

I agree, choose a good seed is cheat.

But I need to make sure this code can work. Is there another continuous action space environment to use, which is very simple?

File: src/algorithms/policy_gradient/vpg.jl (hunk 12 above)
Comment (rbange) [reply src/algorithms/policy_gradient/vpg.jl]: Nice, maybe your [issue](https://github.com/JuliaStats/Distributions.jl/issues/1183) will raise some attention towards that topic.

Im not really comfortable with using this kind of 'hacks' in a frameworks because it basically creates redundant code, but yes it would help a lot. Maybe @findmyway can comment on finding a good place for such patches?

Regarding my implementation:
I defined an constant for `log2π` to avoid Float64 creeping into this, but something like `log(2f0π)` will work too of course!

Comment (rbange) [reply src/algorithms/policy_gradient/vpg.jl]: But for the sake of merging this PR, I think it might be best to fix this in at a later point in a new PR?

Comment (findmyway) [reply src/algorithms/policy_gradient/vpg.jl]: I usually put this kind of patches into [ReinforcementLearningCore.jl/tree/master/src/extensions](https://github.com/JuliaReinforcementLearning/ReinforcementLearningCore.jl/tree/master/src/extensions) and try to merge them into upstream packages when possible.

File: src/experiments/rl_envs.jl (hunk 11 above)
Comment (rbange) [reply src/experiments/rl_envs.jl]: Unfortunately, I have not come across a nice environment for this purpose. Guess, that's the reason why mujoco is so present in this field.

https://github.com/JuliaReinforcementLearning/ReinforcementLearningEnvironments.jl has an implementation of `ContinuousMountainCarEnv`. But that version is implemented with a very sparse reward function which I guess will make converging far more difficult. 

In theory there is also a continuous CartPole version, but I never managed to get an algorithm to converge in this env, not sure what's the issue there.

Finally, you can always fall back to some other Frameworks like OpenAI Gym (e.g. the Box2d environments), but these experiments should not be included in the tests of course due to the additional setup steps required.

**Comment by findmyway** [comment]
I'll merge this first and we can work on the issue discussed https://github.com/JuliaReinforcementLearning/ReinforcementLearningZoo.jl/pull/87#discussion_r493494320 later.

File: src/experiments/rl_envs.jl (hunk 11 above)
Comment (norci) [reply src/experiments/rl_envs.jl]: I tried the continuous version MountainCarEnv, I think it's impossible to converge. I logged the reward in every episode, it rarely got reward. so there's no gradient at all.

I also tried to use PG from https://github.com/thu-ml/tianshou, with Pendulum, it did not converge too.

In fact tianshou did not implement continuous action space case at all. I modified a few code, in order to use Gaussian policy agent.
PARL (https://github.com/PaddlePaddle/PARL/blob/develop/parl/algorithms/torch/policy_gradient.py) only supports Categorical distribution too.

So I guess VPG with a Gaussian policy is not practical.
